### PR TITLE
video: Free framebufer of previous running app

### LINKED ISSUE #357: `AvGetSavedDataAddress` / `AvSetSavedDataAddress` is never free'd or used
`AvGetSavedDataAddress` returns the address of a persistent allocation which contains the framebuffer contents of the previous application (as a D3D surface). At 1080i, this buffer would be about 8MiB. That memory will be unavailable to the application.
This is mostly a documentation issue, because it should probably be the applications responsibility (or that of some API) to read it out / release it.

As a quick workaround, we could potentially integrate this with XVideo by freeing the buffer during `XVideoSetMode`?

### PR DESCRIPTION
**Pull request by kosmas12**
This has been tested to cause no problems on xemu and a 1.1 Xbox. The project I used to test with is [https://github.com/Ryzee119/SDLPoPX](url).

Thanks to @kl0wn for looking into how this can be achieved!

Closes #357

### CODE REVIEW

**Review by thrimbor — CHANGES_REQUESTED** [review]
The indentation is different in comparison to the surrounding code (tabs vs. spaces).

I don't think `XVideoSetMode` is the best place for this - imho it makes more sense to put it into `XVideoInit`, so the framebuffer doesn't get freed when failing to set a new mode. There, we could also compare the size of the old framebuffer to the framebuffer size of the requested mode, and reuse the old framebuffer instead of allocating a new one.

To fully resolve the linked issue, we should also mark newly allocated framebuffers as persistent, and store their address with `AvSetSavedDataAddress`, so other applications can reuse our framebuffer.

**Comment by kosmas12** [comment]
I applied all the suggested improvements and tested them on both xemu and a 1.1 Xbox with [https://github.com/dracc/NevolutionX](url) and [https://github.com/Ryzee119/SDLPoPX](url), all seems to work fine.

**Review by thrimbor — APPROVED** [review]

**Comment by thrimbor** [comment]
I took the liberty of fixing some smaller remaining issues myself (whitespaces, formatting, commit title too long), merging now.

File: lib/hal/video.c [hunk 1]
```diff
@@ -357,7 +363,28 @@ void XVideoInit(DWORD dwMode, int width, int height, int bpp)
 	/* Store the framebuffer that we set; normally this is done in XVideoSetFB.
 	   However, we're using the kernel's AvSetDisplayMode here instead, so we
 	   set it here, too. */
-	_fb = framebufferMemory;
+
+	// Get previous framebuffer
+	PVOID previousFB = AvGetSavedDataAddress();
+
+	if (previousFB != 0) {
```
Comment (GXTX) [inline lib/hal/video.c:370]: Shouldn't this be `NULL`?

File: lib/hal/video.c [hunk 2]
```diff
@@ -357,7 +363,28 @@ void XVideoInit(DWORD dwMode, int width, int height, int bpp)
...
+		SIZE_T previousFBSize = MmQueryAllocationSize(previousFB);
+		// If the previous framebuffer is the same size as the one we want to create, don't create a new one and use that instead
+		if (previousFBSize == screenSize) {
+			memset(previousFB, 0x00, screenSize);
```
Comment (GXTX) [inline lib/hal/video.c:374]: Shouldn't this use `RtlZeroMemory` / `ZeroMemory()`?

File: lib/hal/video.c [hunk 3]
```diff
@@ -357,7 +363,28 @@ void XVideoInit(DWORD dwMode, int width, int height, int bpp)
...
+			_fb = (unsigned char *)previousFB;
+		} else {
+			MmPersistContiguousMemory(previousFB, previousFBSize, FALSE);
+			MmFreeContiguousMemory(previousFB);
+			AvSetSavedDataAddress(NULL);
```
Comment (GXTX) [inline lib/hal/video.c:379]: Is this required? We set a new address once we're out of the if/else block.

**Pull request merged**